### PR TITLE
If we extend the DM table return an indicator to save the fs metadata

### PR DESCRIPTION
**Pull request by trgill**
Still needs testing to address #1474

### CODE REVIEW

**Comment by ingothierack** [comment]
If i can get a package to install I will give it a try. But one question, will this add also the current missing part of the filesystem, or it this lost?

**Comment by mulkieran** [comment]
> If i can get a package to install I will give it a try. But one question, will this add also the current missing part of the filesystem, or it this lost?

We believe you should be able to get back the missing part of the filesystem by editing some Stratis metadata and doing another reboot. @trgill Would you still be able to test if this would work?

**Comment by ingothierack** [comment]
Any short howto to reaktive is welcome. In hard case, i can restore from backup. But if i can't create stable filesystems > 1TB it's useless :(

**Comment by mulkieran** [comment]
@ingothierack 

Here's the short version of what I think is the best approach:

You should find a file with pathname something like:
```/stratis/.mdv-210eb8dac6cb440993454dac6a885dae/filesystems/f4a9283f89c3457bb5980d94eb82cc34.json```.
The first UUID is the pool's, the second the filesystem's.
This contains the JSON metadata for the particular filesystem. Based on your symptoms, I would expect you to see a key/value pair, ```size: 2147483648```. The size is in sectors of 512 B, that's equivalent to 1 TiB. Update this value to 2 TiB, 4294967296, and reboot once more. This _should_ give you the correctly sized thin device, and result in a mountable filesystem.

**Comment by ingothierack** [comment]
i did not have this kind of structure under /stratis
Visible the pools as directory and then under pool the symlinks to the dm-devices
I the the dumps of it in the logfile. but didnt' find the location of if

`[root@nas001 stratis]# ls -lrta
total 4
dr-xr-xr-x. 18 root root 4096 Mar  6 21:46 ..
drwxr-xr-x   3 root root   28 Apr  2 18:33 .
drwxr-xr-x   2 root root  291 Apr  2 18:33 stratis_pool01
[root@nas001 stratis]# cd stratis_pool01/
[root@nas001 stratis_pool01]# ls -lrta
total 0
drwxr-xr-x 3 root root  28 Apr  2 18:33 ..
lrwxrwxrwx 1 root root  10 Apr  2 18:33 fs_workarea -> /dev/dm-30
lrwxrwxrwx 1 root root  10 Apr  2 18:33 fs_torrent -> /dev/dm-22
lrwxrwxrwx 1 root root  10 Apr  2 18:33 fs_stage -> /dev/dm-36
lrwxrwxrwx 1 root root  10 Apr  2 18:33 fs_share01 -> /dev/dm-37
lrwxrwxrwx 1 root root  10 Apr  2 18:33 fs_remotebackup -> /dev/dm-27
lrwxrwxrwx 1 root root  10 Apr  2 18:33 fs_picture -> /dev/dm-26
lrwxrwxrwx 1 root root  10 Apr  2 18:33 fs_mirror -> /dev/dm-35
lrwxrwxrwx 1 root root  10 Apr  2 18:33 fs_media -> /dev/dm-24
lrwxrwxrwx 1 root root  10 Apr  2 18:33 fs_home -> /dev/dm-32
lrwxrwxrwx 1 root root  10 Apr  2 18:33 fs_git -> /dev/dm-34
lrwxrwxrwx 1 root root  10 Apr  2 18:33 fs_download -> /dev/dm-25
lrwxrwxrwx 1 root root  10 Apr  2 18:33 fs_dockerrepo -> /dev/dm-28
lrwxrwxrwx 1 root root  10 Apr  2 18:33 fs_bucketstore -> /dev/dm-29
lrwxrwxrwx 1 root root  10 Apr  2 18:33 fs_book -> /dev/dm-23
lrwxrwxrwx 1 root root  10 Apr  2 18:33 fs_backup -> /dev/dm-33
lrwxrwxrwx 1 root root  10 Apr  2 18:33 fs_archiv -> /dev/dm-31
drwxr-xr-x 2 root root 291 Apr  2 18:33 .`

**Comment by mulkieran** [comment]
> i did not have this kind of structure under /stratis
> Visible the pools as directory and then under pool the symlinks to the dm-devices
> I the the dumps of it in the logfile. but didnt' find the location of if
> 
> ```
> total 4
> dr-xr-xr-x. 18 root root 4096 Mar  6 21:46 ..
> drwxr-xr-x   3 root root   28 Apr  2 18:33 .
> drwxr-xr-x   2 root root  291 Apr  2 18:33 stratis_pool01
> [root@nas001 stratis]# cd stratis_pool01/
> [root@nas001 stratis_pool01]# ls -lrta
> total 0
> drwxr-xr-x 3 root root  28 Apr  2 18:33 ..
> lrwxrwxrwx 1 root root  10 Apr  2 18:33 fs_workarea -> /dev/dm-30
> lrwxrwxrwx 1 root root  10 Apr  2 18:33 fs_torrent -> /dev/dm-22
> lrwxrwxrwx 1 root root  10 Apr  2 18:33 fs_stage -> /dev/dm-36
> lrwxrwxrwx 1 root root  10 Apr  2 18:33 fs_share01 -> /dev/dm-37
> lrwxrwxrwx 1 root root  10 Apr  2 18:33 fs_remotebackup -> /dev/dm-27
> lrwxrwxrwx 1 root root  10 Apr  2 18:33 fs_picture -> /dev/dm-26
> lrwxrwxrwx 1 root root  10 Apr  2 18:33 fs_mirror -> /dev/dm-35
> lrwxrwxrwx 1 root root  10 Apr  2 18:33 fs_media -> /dev/dm-24
> lrwxrwxrwx 1 root root  10 Apr  2 18:33 fs_home -> /dev/dm-32
> lrwxrwxrwx 1 root root  10 Apr  2 18:33 fs_git -> /dev/dm-34
> lrwxrwxrwx 1 root root  10 Apr  2 18:33 fs_download -> /dev/dm-25
> lrwxrwxrwx 1 root root  10 Apr  2 18:33 fs_dockerrepo -> /dev/dm-28
> lrwxrwxrwx 1 root root  10 Apr  2 18:33 fs_bucketstore -> /dev/dm-29
> lrwxrwxrwx 1 root root  10 Apr  2 18:33 fs_book -> /dev/dm-23
> lrwxrwxrwx 1 root root  10 Apr  2 18:33 fs_backup -> /dev/dm-33
> lrwxrwxrwx 1 root root  10 Apr  2 18:33 fs_archiv -> /dev/dm-31
> drwxr-xr-x 2 root root 291 Apr  2 18:33 .```
> ```

I overlooked: for this approach, you will have to mount the mdv (metadata volume) in order to edit that metadata. For that you will need to locate the mdv device.

The other approach is to use dmsetup to update the thindev, and then trick stratisd into properly recording the change. I will see how that would work.

**Comment by ingothierack** [comment]
got it work by mounting the metadata volume and edit the file.  Found the correct dev/dm-xx device.
After the reboot as you suggested could mount the filesystem.
If this is now stable, i think, i can wait until v1.1 if there is for this the fix included.

**Comment by mulkieran** [comment]
@ingothierack Thanks for being so resourceful! The current size of the filesystem will not be forgotten. However, if you extend the filesystem again, beyond the current 2TiB limit, the new size will not be properly recorded in the metadata. This PR should fix that problem.

**Comment by ingothierack** [comment]
@mulkieran thats ok for me. No current filesystem will be grow beyond 1 TB or > 2TB in a short time, hopefully. If I hit this, i could raise the limit also manually. But this is not for the mainstream. So the fixes should be in a Release 1.0.4 or we see in a very short time 1.1. Personally i would like to see in the first time a few more updates here. Later we can see a lower speed in updates, feature wise, but bugfixes  should be released immediately, even we got a lot of releases. I thinks, 1.0 released to early for what i see now.

Also thanks for you suggestion to fix this.

File: src/engine/strat_engine/thinpool/thinpool.rs [hunk 1]
```diff
@@ -553,18 +552,16 @@ impl ThinPool {
...
-            .iter_mut()
-            .map(|(_, _, fs)| fs.check())
-            .collect::<StratisResult<Vec<_>>>()?;
-
-        for fs_status in filesystems {
+        for (name, uuid, fs) in self.filesystems.iter_mut() {
+            let (fs_status, save_mdv) = fs.check()?;
+            if save_mdv && self.mdv.save_fs(name, uuid, &fs).is_err() {
+                error!("Failed to save MDV for : {}", uuid);
+            }
```
Comment (trgill) [inline src/engine/strat_engine/thinpool/thinpool.rs]: @mulkieran any suggestions for doing this in a more idiomatic way?  I couldn't get it working with the closure.

**Comment by trgill** [comment]
Adding a test that would have caught the original problem - not ready for review.

**Comment by trgill** [comment]
@mulkieran I added the test to verify thin dev extension after reconstruction

**Comment by mulkieran** [comment]
@trgill Something very strange w/ stratisd Jenkins test right now. I get requested for a password multiple times, but I do proceed further and the log is so long that it can not be displayed.

**Comment by trgill** [comment]
@mulkieran I'm looking at the CI results now.  I added the "do not merge" tag back until we understand what is happening.

**Comment by trgill** [comment]
The CI failures can be explained by : https://github.com/stratis-storage/stratisd/issues/1446

@mulkieran please review?

**Pull request closed without merging**